### ISSUE TITLE
GC overhead limit exceeded on opening large .GMK

### ISSUE THREAD

**Issue by BPzeBanshee** [IsmAvatar/LateralGM#450]
Full error message on Pastebin: https://pastebin.com/AUrQKFxx
Project file in question: https://drive.google.com/open?id=1Dftdwk0qhWY1XkkzIDcgKnl4QAeu-_pa

Not gonna lie, it's a huge unoptimised project and I'm basically just posting on behalf of the project maintainer to see if anything can be done. It seems whatever method LateralGM uses to open this up into memory hits a hard limit. I've tried using the heap size command arguments to no effect on my end.

**Comment by RobertBColton** [comment]
Hey! Let me say I can't reproduce, but I do believe you. I got it to load myself and it seemed to take the longest on rooms. This is LGM's ongoing weakness compared to RGM which we were working on. LGM loads the whole project into memory. Again, that's why GMS forces you to convert GMKs to GMX. It's hard to stream a GMK in chunks because it's just a giant binary blob.

**Comment by BPzeBanshee** [comment]
Thanks for the heads up Robert. I'll take a look at my environment and see if I can make it work with JDK9. Any particular commandlines you use when launching it (so I can rule out screwing up an argument)?

**Comment by RobertBColton** [comment]
No, nothing special, I'm cheating a bit too because I have 16 GB ram. I can tell you my exact Java though if it helps.

```
$ java -version
java version "9.0.4"
Java(TM) SE Runtime Environment (build 9.0.4+11)
Java HotSpot(TM) 64-Bit Server VM (build 9.0.4+11, mixed mode)
```

**Comment by RobertBColton** [comment]
I have found the bottleneck, it's ActiveArrayList in the Room resource. The room uses these to keep track of the tiles and instances.
https://github.com/IsmAvatar/LateralGM/blob/918dff9cb922cef656a91d2aec0ad266eea8b79e/org/lateralgm/resources/Room.java#L44

If I keep the GMK reader the same, still loading tiles and instances, but simply comment out adding the read tiles and instances to these lists, it stops LGM from exploding. Your game then loads fine. Some deeper inspection of why ActiveArrayList is performing subpar is needed.

**Comment by RobertBColton** [comment]
I collected this from Memory Analyzer which comes with Eclipse, it seems to suspect the same thing I do, but I need Josh or Rusky to understand it.

![Eclipse Memory Analyzer](https://user-images.githubusercontent.com/3212801/65814273-9a8c7980-e1ad-11e9-89b7-d2bea5ff1f42.png)

**Comment by JoshDreamland** [comment]
Those errors seem to suggest that the game contains over four million resources, and that a single tile is occupying 300MB. My guess is GMK corruption, or a glitch in the reader. Assuming the entire GMK is less than 300MB.

**Comment by RobertBColton** [comment]
Way off Josh, it's 155,000 tiles equaling 300mb of RAM. The question is why so few tiles is so big, they don't take 300mb.

After some more digging, I realized the update source stuff for each property is causing about half the bloat. Let me clarify that adding the tiles/instances to the room array lists just meant they were being retained rather than garbage collected, so not the real issue. Resetting my clone and then commenting out the update source lines in the constructor for Tile/Instance makes the game load in at 2gb. This seems to be the real source of contention.
https://github.com/IsmAvatar/LateralGM/blob/918dff9cb922cef656a91d2aec0ad266eea8b79e/org/lateralgm/resources/sub/Instance.java#L57

**Comment by RobertBColton** [comment]
Ok, so let me reiterate what's wrong here so people understand. In Java, every object has a lock for synchronization. This means that every object in Java is usually at least 20 bytes~ big. This means an int, which is normally 8 bytes, becomes 20 bytes if you box it inside the Integer class. LGM uses a property map for each of its resource types, meaning every single property of a resource (such as tile or instance) is at least 20 bytes. If that isn't bad enough, LGM then uses an update listener with a pair of objects for each property, adding another 40~ bytes per property.

The problem here is not just that Java is arguably a poor language. LateralGM was also simply engineered incorrectly since it did not consider the implications of any of this during its design. Multiple problems here and all of this would require a significant amount of engineering to fix.

**Comment by BPzeBanshee** [comment]
Makes sense to me. I guess adding a WontFix label would be reasonable here given the intent to move away from LateralGM for future ENIGMA IDE use.

A few more questions if you don't mind:
- How is the state of RadialGM in this respect? Could the kind of engineering required to solve this problem be implemented better over in that project instead?
- How does one defrag IDs in GM 8.1? I can't find the option anywhere.

**Comment by RobertBColton** [comment]
Basically, yeah. For RadialGM, this problem is pretty much already solved. First, libEGM is C++ and Google Protocol Buffers based, so each tile/property are using a normal amount of space you would expect for the fields. An int32 field is exactly 8 bytes, no extra lock, no boxing, etc. Then RadialGM loads each individual resource into a data model which is just an 8 byte pointer. And it can still hook up the editors to reflect changes in the properties of each other without taking 40 bytes per property! Also, because RGM wasn't focusing on writing GM formats back out, only importing and using EGM, the idea was to lazy load or stream large blobs like images and sounds.

Again, not completely deriding Java. LGM could have worked better with a lot of TLC, but it's wasted when RGM really is such a good idea.
https://ordepdev.me/posts/size-of-an-object-in-java

To defrag IDs in GM8.1 now that you mention it, I think GM8.1 will only ask you if it detects duplicate ids, which I see the project doesn't have. That doesn't mean the ids couldn't be compressed though. As you know, ids are assigned linearly by LGM and old GM, from 1 to INT_MAX or whatever. Each time you delete a resource, it's actually still kept, there's an extra boolean in the GMK telling the reader to skip over it, but the high id remains the same. When the ids are defragged it clears all the deleted resource gaps and makes your resources flat again. This decreases the size of the GMK. However, I don't know now if this will actually improve LGM with it, because instances/tiles do not work this way.

**Comment by RobertBColton** [comment]
Understanding this limitation, I decided to do #457 and allow LGM users to forcefully defrag even if ids are already unique. You were previously capable of doing this in LGM prior to me making the menu option interactive and having it check the ids first. I attempted to test this on the project above myself, but it never completed after about 15 minutes of letting it run. I did test it on some other largish GMKs, like Bowser's Last Stand. For that game, the force defrag effectively reduced the GMK file size by 0.01 mbs.

This is a step in the right direction, if not solving this issue yet.

https://github.com/IsmAvatar/LateralGM/releases/tag/v1.8.89

**Comment by RobertBColton** [comment]
A very small part of this problem has been fixed. With 72844ab2e2141742701a40c642151eb7afb535ba I added caching of the lookup of all of the icons which was really poor performance before because we were reading the entire icon key map every single time for every single icon. Now when your project is opened in the latest 1.8.150, it still uses 4gb RAM for me, however, it's at least navigable as in I can open the tree folders and resources without much more lag than a normal size project.

So the rest of this issue is just really figuring out a way to make LGM use less memory for its resources/specifically PropertyMap.

**Comment by BPzeBanshee** [comment]
Thanks for the updates Robert, I appreciate you trying to keep me posted. :)

Dooki (the main project lead for this file) has been doing some work to try and optimise things on their end. They've pushed the project up to GM 8.1 with the intent of eventually migrating it to Studio and getting around the issues of older GM's fragmented ID handling completely. Perhaps with this change combined with the optimisations on their end it might actually be openable on LateralGM now, despite the problems with PropertyMap still being present.

**Comment by RobertBColton** [comment]
@BPzeBanshee The file was taken down, can you bring it back up? I've sent #543 and I'd like to test it on it.

**Comment by BPzeBanshee** [comment]
I'll try and locate it after work. The project in question has gone through a few revisions now including some GMKSplitter rebuilds thanks to the issues discovered indirectly from this ticket, so I don't know offhand if the newer file is the same size.On 6 Jan. 2021 4:34 am, Robert Colton <notifications@github.com> wrote:
@BPzeBanshee The file was taken down, can you bring it back up? I've sent #543 and I'd like to test it on it.

**Comment by BPzeBanshee** [comment]
Looks like they've just had a recent update with source available from this page: https://fusion-fangaming.itch.io/mushroom-kingdom-fusion

If you *really* want to stress test LateralGM with this beast of a project, there's links from that site to Google Drives with older & less optimised GMKs.

**Comment by RobertBColton** [comment]
Thanks, I was using Iji temporarily to test, but this one is much better because of how big it is. So, I tested again v1.8.210 (before #543 was merged) and after v1.8.211, and it seems it was the right decision. The old jar loads at 4241 MB RAM and takes 19,453 ms, while the new jar used 3003 MB RAM and took 17,888 ms to load. I actually did not expect a change in performance from that, but the decrease in memory consumption makes it clear it was the right decision.

The only thing left here to get LGM to 900~ MB RAM, like GM8.1, would be to get rid of PropertyMap and have the resources actually use int, float, double for its properties. To get closer to C like cache performance, we could also do structure of arrays or array of structures (for which there are already libraries in Java). That would still be quite a bit of work, but it's possibly doable.

PS: Even Windows Explorer lagged heavily when I browsed the project data folders. These guys are definitely testing the bounds here with GM in general!
